### ISSUE TITLE
t0001-basic.t: not ok 17 - and broker.boot-method=single

### ISSUE THREAD

**Issue by chu11** [flux-framework/flux-core#4873]
this is on catalyst

```
not ok 17 - and broker.boot-method=single
#	
#		test libpmi = single
#	
```
appears to have started after PR #4865 was merged.

**Comment by grondo** [comment]
Is this the test from `t0001-basic.t`? (If so, let's add that to the issue description)

However, this doesn't reproduce for me on catalyst. Can you try this in your environment?
```console
$ src/cmd/flux start -o,-Sbroker.rc1_path=,-Sbroker.rc3_path= flux getattr broker.boot-method
single
```

**Comment by chu11** [comment]
oops yeah, was `t0001-basic.t`.

```
>src/cmd/flux start -o,-Sbroker.rc1_path=,-Sbroker.rc3_path= flux getattr broker.boot-method
libpmi
```

could be some weird pathing thing?

**Comment by grondo** [comment]
Might be helpful to try with `FLUX_PMI_DEBUG=1` set, e.g.:
```console
$ FLUX_PMI_DEBUG=1 src/cmd/flux start -o,-Sbroker.rc1_path=,-Sbroker.rc3_path= flux getattr broker.boot-method
boot_pmi: flux-pmi-client: trying 'simple'
boot_pmi: flux-pmi-client: trying 'libpmi'
boot_pmi: flux-pmi-client: trying 'single'
boot_pmi: single: initialize: rank=0 size=1 name=single: success
boot_pmi: single: get key=flux.instance-level: key not found
boot_pmi: single: finalize: successs
single
```

**Comment by garlick** [comment]
Haven't tried yet, but if slurm's libpmi.so is installed in ${libdir} it might be hitting that.
Perhaps this test should set FLUX_PMI_CLIENT_SEARCHPATH="simple single" to prevent that...

**Comment by chu11** [comment]
it does seem to be finding `libpmi` somewhere.

```
>FLUX_PMI_DEBUG=1 src/cmd/flux start -o,-Sbroker.rc1_path=,-Sbroker.rc3_path= flux getattr broker.boot-method
boot_pmi: flux-pmi-client: trying 'simple'
boot_pmi: flux-pmi-client: trying 'libpmi'
boot_pmi: libpmi: initialize: rank=0 size=1 name=0.0: success
boot_pmi: libpmi: get key=flux.instance-level: invalid key argument
boot_pmi: libpmi: finalize: successs
libpmi
```

slurms libpmi is definitely on catalyst, although not clear why i'm hitting it but @grondo didn't.  unloaded modules, tried setting `.notce`, but not an obvious thing that my environment could be doing that is strangely different.

@garlick  I think you meant FLUX_PMI_CLIENT_METHODS, setting that did work

```
>FLUX_PMI_CLIENT_METHODS="simple single" FLUX_PMI_DEBUG=1 src/cmd/flux start -o,-Sbroker.rc1_path=,-Sbroker.rc3_path= flux getattr broker.boot-method
boot_pmi: flux-pmi-client: trying 'simple'
boot_pmi: flux-pmi-client: trying 'single'
boot_pmi: single: initialize: rank=0 size=1 name=single: success
boot_pmi: single: get key=flux.instance-level: key not found
boot_pmi: single: finalize: successs
single
```

**Comment by garlick** [comment]
Hmm, I can't reproduce it either on catalyst160.   We do find a `libpmi.so` and it must be slurm's:
```
[garlick@catalyst160:flux-core]$ ls -la /usr/lib64/libpmi.so
lrwxrwxrwx 1 root root 15 Jan 18 01:36 /usr/lib64/libpmi.so -> libpmi.so.0.0.0
[garlick@catalyst160:flux-core]$ rpm -qf /usr/lib64/libpmi.so.0.0.0
slurm-libpmi-21.08.8-2.1.ch6.x86_64
```
And it's failing in `PMI_Init()`, which is different than before with slurm.  Slurm's PMI used to fall back to singleton operation if it didn't find itself in a slurm job.

@chu11 is your `usr/lib64/libpmi.so.0.0.0` the same as the one on catalyst160?    Is it possible you are running the test from within a slurm job?  (I tried an salloc but it didn't repro for me there either)

Regardless, I think the test is fragile and not doing very much, and we could just drop it.

**Comment by grondo** [comment]
In case it is helpful, I added some debugging on catalyst and see this in my environment:

```
$ FLUX_PMI_DEBUG=1 src/cmd/flux start -o,-Sbroker.rc1_path=,-Sbroker.rc3_path= flux getattr broker.boot-method
boot_pmi: flux-pmi-client: trying 'simple'
boot_pmi: simple: simple: PMI_FD not found in environ
boot_pmi: flux-pmi-client: trying 'libpmi'
boot_pmi: libpmi: libpmi: libpmi.so: dlopen found Flux library (/g/g0/grondo/git/flux-core.git/src/common/.libs/libpmi.so)
boot_pmi: flux-pmi-client: trying 'single'
boot_pmi: single: initialize: rank=0 size=1 name=single: success
boot_pmi: single: get key=flux.instance-level: key not found
boot_pmi: single: finalize: successs
single
```

**Comment by grondo** [comment]
FWIW, printing the error after each boot method test failure could be valuable in other situations, and I also note there's a typo `successs` that could be fixed in any PR addressing this issue :slightly_smiling_face:

**Comment by garlick** [comment]
Excellent!  I was really confused there.

The source tree libpmi.so would have been left over from a previous build as it's now located in a flux subdirectory to prevent this sort of thing.  Oh look I have one too and when I remove it, @chu11's observation reproduces:
```
[garlick@catalyst160:flux-core]$ find . -name libpmi.so
./src/common/.libs/libpmi.so
./src/common/flux/.libs/libpmi.so
[garlick@catalyst160:flux-core]$ rm -f src/common/.libs/libpmi.so
[garlick@catalyst160:flux-core]$ FLUX_PMI_DEBUG=1 src/cmd/flux start -o,-Sbroker.rc1_path=,-Sbroker.rc3_path= flux getattr broker.boot-method
boot_pmi: flux-pmi-client: trying 'simple'
boot_pmi: flux-pmi-client: trying 'libpmi'
boot_pmi: libpmi: initialize: rank=0 size=1 name=0.0: success
boot_pmi: libpmi: get key=flux.instance-level: invalid key argument
boot_pmi: libpmi: finalize: successs
libpmi
```

**Comment by grondo** [comment]
Oh my. Sorry, I should have tried from a fresh clone! Thank goodness @chu11 has good working tree hygiene!

**Comment by garlick** [comment]
@grondo how did you retrieve the path of the library that `dlopen()` found?  I was wanting to do that earlier but didn't see a way.

**Comment by grondo** [comment]
This was the debug I was using:
```diff
diff --git a/src/common/libpmi/upmi.c b/src/common/libpmi/upmi.c
index 4835a6888..d7411cd05 100644
--- a/src/common/libpmi/upmi.c
+++ b/src/common/libpmi/upmi.c
@@ -270,6 +270,7 @@ struct upmi *upmi_create (const char *uri,
                 && upmi_preinit (upmi, upmi->flags, NULL, error) == 0) {
                 break;
             }
+            upmi_trace (upmi, "%s: %s", uri, error->text);
             upmi->plugin = NULL;
             uri = argz_next (upmi->methods, upmi->methods_len, uri);
         }
diff --git a/src/common/libpmi/upmi_libpmi.c b/src/common/libpmi/upmi_libpmi.c
index d0e12dfd2..16b76cd79 100644
--- a/src/common/libpmi/upmi_libpmi.c
+++ b/src/common/libpmi/upmi_libpmi.c
@@ -13,8 +13,11 @@
 #endif
 #include <stdlib.h>
 #include <errno.h>
+
+#include <link.h>
 #include <dlfcn.h>
 
+
 #include "src/common/libutil/errprintf.h"
 #include "ccan/array_size/array_size.h"
 
@@ -47,6 +50,7 @@ static int dlopen_wrap (const char *path,
                         flux_error_t *error)
 {
     void *dso;
+    struct link_map *p;
 
     dlerror ();
     if (!(dso = dlopen (path, flags))) {
@@ -57,9 +61,10 @@ static int dlopen_wrap (const char *path,
             errprintf (error, "%s: dlopen failed", path);
         return -1;
     }
+    dlinfo (dso, RTLD_DI_LINKMAP, &p);
     if (noflux) {
         if (dlsym (dso, "flux_pmi_library") != NULL) {
-            errprintf (error, "%s: dlopen found Flux library", path);
+            errprintf (error, "%s: dlopen found Flux library (%s)", path, p->l_name);
             goto error;
         }
     }
```

**Comment by garlick** [comment]
closed by #4876
